### PR TITLE
Remove `parking_lot` dependency

### PR DESCRIPTION
**Pull request by madsmtm**
Fixes https://github.com/rust-windowing/winit/issues/650.

`std`'s `Mutex` has [recently improved a lot](https://blog.rust-lang.org/2022/06/30/Rust-1.62.0.html#thinner-faster-mutexes-on-linux), so there's not really a performance argument for keeping i any more. Also means we won't have to use the same `windows-sys` version as `parking_lot`

Dependency tree (Windows):
```
// Before
winit v0.27.1
├── bitflags v1.3.2
├── instant v0.1.12
│   └── cfg-if v1.0.0
├── log v0.4.17
│   └── cfg-if v1.0.0
├── once_cell v1.13.0
├── parking_lot v0.12.1
│   ├── lock_api v0.4.7
│   │   └── scopeguard v1.1.0
│   │   [build-dependencies]
│   │   └── autocfg v1.1.0
│   └── parking_lot_core v0.9.3
│       ├── cfg-if v1.0.0
│       ├── smallvec v1.9.0
│       └── windows-sys v0.36.1
│           └── windows_x86_64_gnu v0.36.1
├── raw-window-handle v0.5.0
│   └── cty v0.2.2
└── windows-sys v0.36.1 (*)

// After
winit v0.27.1
├── bitflags v1.3.2
├── instant v0.1.12
│   └── cfg-if v1.0.0
├── log v0.4.17
│   └── cfg-if v1.0.0
├── once_cell v1.13.0
├── raw-window-handle v0.5.0
│   └── cty v0.2.2
└── windows-sys v0.36.1
    └── windows_x86_64_gnu v0.36.1
```

Dependency tree (X11):
```
// Before
winit v0.27.1
├── bitflags v1.3.2
├── instant v0.1.12
│   └── cfg-if v1.0.0
├── libc v0.2.129
├── log v0.4.17
│   └── cfg-if v1.0.0
├── mio v0.8.4
│   ├── libc v0.2.129
│   └── log v0.4.17 (*)
├── once_cell v1.13.0
├── parking_lot v0.12.1
│   ├── lock_api v0.4.7
│   │   └── scopeguard v1.1.0
│   │   [build-dependencies]
│   │   └── autocfg v1.1.0
│   └── parking_lot_core v0.9.3
│       ├── cfg-if v1.0.0
│       ├── libc v0.2.129
│       └── smallvec v1.9.0
├── percent-encoding v2.1.0
├── raw-window-handle v0.5.0
│   └── cty v0.2.2
└── x11-dl v2.20.0
    ├── lazy_static v1.4.0
    └── libc v0.2.129
    [build-dependencies]
    └── pkg-config v0.3.25

// After
winit v0.27.1
├── bitflags v1.3.2
├── instant v0.1.12
│   └── cfg-if v1.0.0
├── libc v0.2.129
├── log v0.4.17
│   └── cfg-if v1.0.0
├── mio v0.8.4
│   ├── libc v0.2.129
│   └── log v0.4.17 (*)
├── once_cell v1.13.0
├── percent-encoding v2.1.0
├── raw-window-handle v0.5.0
│   └── cty v0.2.2
└── x11-dl v2.20.0
    ├── lazy_static v1.4.0
    └── libc v0.2.129
    [build-dependencies]
    └── pkg-config v0.3.25
```

- [ ] Tested on all platforms changed
  - Nope, please test this!
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

File: src/platform/unix.rs
Comment (notgull) [inline src/platform/unix.rs:59]: While this is entirely fine, I'd suggest having a macro like this somewhere in the crate:

```rust
macro_rules! lock {
    ($mutex: expr) => {{
        match ($mutex).lock() {
            Ok(guard) => guard,
            Err(err) => {
                log::warn!("Mutex is poisoned: {:?}", &err);
                err.into_inner()
            }
        }
    }}
}
```

And instead calling it like this:

```rust
lock!(XLIB_ERROR_HOOKS).push(hook);
```

This ensures two things:

1. A poisoned lock won't blow up the entire system.
2. If we decide to switch to another mutex scheme, for some reason, we can easily adjust `lock!` to fit our use case.

**Review by notgull — APPROVED** [review]
This looks good to me; I'll always support removing a dependency that's been obsoleted by the standard library.

File: src/platform/unix.rs
Comment (madsmtm) [reply src/platform/unix.rs:59]: I went with `unwrap` because it's what's used in the other backends, and I'm not at all sure how to handle mutex poisoning

Comment (notgull) [reply src/platform/unix.rs:59]: Hmm, I guess I should add it in a separate PR then...

Comment (madsmtm) [reply src/platform/unix.rs:59]: That would be nice, please do!

Comment (notgull) [reply src/platform/unix.rs:59]: Great! I'll wait until after this PR is merged, so I can make a macro that works across the entire crate.

**Review by msiglreith — APPROVED** [review]

File: src/platform/unix.rs
Comment (madsmtm) [reply src/platform/unix.rs:59]: Hey @notgull, this PR has been merged, so if you want to make this change, now you have the opportunity!

**Pull request merged**
